src/main/perl/Check.pm

```diff
@@ -0,0 +1,524 @@
+    $self->_reset_exception_fail();
+
+    return SUCCESS if (! $self->any_exists($dest));
+
+    $backup = $self->{backup} if (! defined($backup));
+
+    # old is the backup location or undef if no backup is defined
+    # (empty string as backup is not allowed, but 0 is)
+    # 'if ($old)' can safely be used to test if a backup is needed
+    my $old;
+    $old = $dest.$backup if (defined($backup) and $backup ne '');
+
+    # cleanup previous backup, no backup of previous backup!
+    my $method;
+    my @args = ($dest);
+    if ($old) {
+        if (! $self->cleanup($old, '', %opts)) {
+            return $self->fail("cleanup of previous backup $old failed");
+        };
+
+        # simply rename/move dest to backup
+        # works for files and directories
+        $method = 'move';
+        push(@args, $old);
+    } else {
+        if($self->directory_exists($dest)) {
+            $method = 'rmtree';
+        } else {
+            $method = 'unlink';
+        }
+    }
+
+    if($self->_get_noaction($opts{$KEEPS_STATE}, "cleanup: ")) {
+        $self->verbose("NoAction set, not going to $method with args ", join(',', @args));
+    } else {
+        my $res = $self->_safe_eval(
+            $CLEANUP_DISPATCH{$method}, \@args, undef,
+            "Cleanup $method failed to remove $dest",
+            "Cleanup $method removed $dest",
+            );
+        # move and unlink return 0 on failure, set $!
+        # rmtree dies on failure
+        if ($method eq 'rmtree') {
+            return if defined($self->{fail});
+        } else {
+            return $self->fail("Cleanup $method failed to remove $dest: $!") if ! $res;
+        };
+    };
+
+    return CHANGED;
+}
+
+
+=item directory
+
+Make sure a directory exists with proper options.
+
+If the directory does not exists (or the C<temp> option is set),
+it is created (including the parent directories as needed),
+and uses C<LC::Check::directory> via C<LC_Check>.
+
+Returns CHANGED if a change was made, SUCCESS if no changes were made
+and undef in case of failure (and the C<fail> attribute is set).
+
+The return value in absence of failure is a dualvar with integer value
+SUCCESS/CHANGED, and the directory as string value
+(in particular relevant for temporary directories).
+
+Additional options
+
+=over
+
+=item owner/group/mode/mtime : options for C<CAF::Check::status>
+
+=item temp
+
+A boolean if true will create a a temporary directory using
+L<File::Temp::tempdir>.
+
+The directory name is the template to use (any trailing
+C<X> characters will be replaced with random characters by C<tempdir>;
+and the directory name will be padded up to at least 4 C<X>).
+
+The C<CLEANUP> option is also set (an removal
+attempt (incl. any files and/or subdirectries)
+will be made at the end of the program).
+
+=back
+
+=cut
+
+# Differences with LC::Check::directory
+#    only accepts one directory
+#    returns directory name or undef (LC::Check::directory returns number of created directories)
```

Comment:
Re-commenting here as GH seems to have lost it. 

returns a dualvar (containing the directory name) or undef